File: test/create_tests.py

import glob
import numpy
from astropy import wcs
from astropy.io import fits

JS_TEST_TEMPLATE = """
  '%(projection)s': function(test) {
    var header, w, world;
    
    header = fs.readFileSync(path.join(__dirname, 'data', '1904-66_%(projection)s'), 'utf8');
    w = new wcs();
    w.init(header);
    
    world = w.pix2sky(%(x0)d, %(y0)d);
    test.equal(world[0], %(ra0).14f);
    test.equal(world[1], %(dec0).14f);
    
    world = w.pix2sky(%(x1)d, %(y1)d);
    test.equal(world[0], %(ra1).14f);
    test.equal(world[1], %(dec1).14f);
    
    world = w.pix2sky(%(x2)d, %(y2)d);
    test.equal(world[0], %(ra2).14f);
    test.equal(world[1], %(dec2).14f);
    
    test.done();
  },
"""


def get_projection(filename):
    return filename[13:16]

def create_tests():
    for f in glob.glob("data/*.fits.gz"):
        header = fits.getheader(f)
        w = wcs.WCS(header)
        
        pixcrd = numpy.array([[0, 0], [24, 38], [45, 98]], numpy.float_)
        world = w.all_pix2world(pixcrd, 0)
        
        template_obj = {
            "projection": get_projection(f),
        }
        
        for index, value in enumerate(pixcrd):
            template_obj["x%d" % index] = value[0]
            template_obj["y%d" % index] = value[1]
            template_obj["ra%d" % index] = world[index][0]
            template_obj["dec%d" % index] = world[index][1]
        
        template = JS_TEST_TEMPLATE % template_obj
        print template

if __name__ == '__main__':
    create_tests()

File: test/create_tests.c
#include <stdio.h>
#include <string.h>
#include "fitsio.h"
#include <wcslib/wcshdr.h>
#include <wcslib/wcs.h>


int main(void) {
  
  fitsfile *fptr;
  int nkeyrec, nreject, nwcs, status = 0;
  char *header;
  char filepath[27];
  struct wcsprm *wcs;
  
  int nprojections = 28;
  char *projections[] = {"AIR","AIT","ARC","AZP","BON","CAR","CEA","COD","COE","COO","COP","CSC","CYP","HPX","MER","MOL","NCP","PAR","PCO","QSC","SFL","SIN","STG","SZP","TAN","TSC","ZEA","ZPN"};
  
  int pixcrd1[] = {0, 0, 24, 38, 45, 98};
  double pixcrd[] = {0, 0, 24, 38, 45, 98};
  double imgcrd[6], phi[6], theta[6], world[6];
  int stat[6];
  
  for (int i = 0; i < nprojections; i++) {
    
    // Get header from FITS file
    strcpy(filepath, "test/data/1904-66_");
    strcat(filepath, projections[i]);
    strcat(filepath, ".fits");
    
    fits_open_image(&fptr, filepath, READONLY, &status);
    fits_hdr2str(fptr, 1, NULL, 0, &header, &nkeyrec, &status);
    
    wcspih(header, nkeyrec, 0, 0, &nreject, &nwcs, &wcs);
    wcsp2s(wcs, 3, 2, pixcrd, imgcrd, phi, theta, world, stat);
    
    printf("'%s': function(test) {\n", projections[i]);
    printf("  var header, w, world, pixel;\n");
    printf("\n");
    printf("  header = fs.readFileSync(path.join(__dirname, 'data', '1904-66_%s'), 'utf8');\n", projections[i]);
    printf("  w = new wcs();\n");
    printf("  w.init(header);\n");
    printf("\n");
    
    for (int j = 0; j < 6; j += 2) {
      printf("  world = w.pix2sky(%d, %d);\n", pixcrd1[j], pixcrd1[j+1]);
      printf("  test.equal(world[0].toFixed(6), %.6f);\n", world[j]);
      printf("  test.equal(world[1].toFixed(6), %.6f);\n", world[j+1]);
      printf("\n");
      printf("  pixel = w.sky2pix(%f, %f);\n", world[j], world[j+1]);
      printf("  test.equal(Math.round(pixel[0]), %d);\n", pixcrd1[j]);
      printf("  test.equal(Math.round(pixel[1]), %d);\n", pixcrd1[j+1]);
      
      printf("\n");
    }
    
    printf("  test.done();\n");
    printf("},\n\n");
    
    fits_close_file(fptr, &status);
  }
  
  // fits_open_image(&fptr, "examples/m101.fits", READONLY, &status);
  // fits_hdr2str(fptr, 1, NULL, 0, &header, &nkeyrec, &status);
  // wcspih(header, nkeyrec, 0, 0, &nreject, &nwcs, &wcs);
  // wcsp2s(wcs, 3, 2, pixcrd, imgcrd, phi, theta, world, stat);
  // 
  // for (int j = 0; j < 6; j += 2) {
  //   printf("%f, %f\n", world[j], world[j+1]);
  // }
  // 
  // fits_close_file(fptr, &status);
  
  return 0;
}